### PR TITLE
Feature/storage provider upgrade to lighthouse

### PR DESCRIPTION
**Pull request by xlassix**
This Pull Request implements the necessary changes to configure Lighthouse's storage to work with CloudFlare R2. 
Here are the critical changes made:
- Implemented necessary configuration changes to enable seamless storage with CloudFlare R2 and  Lighthouse's storage

### CODE REVIEW

File: lib/poda_payload.py [hunk 1]
```diff
@@ -80,43 +99,60 @@ def send_blobs(self, syscoind):
...
-                                print("checking PoDA txid {0} {1}".format(txid, self.bucketname))
-                                self.s3.Object(self.bucketname, blobresponse.get('versionhash')).load()
+                                print("checking PoDA txid {0} {1}".format(
+                                    txid, self.bucketname))
+                                self.s3.Object(self.bucketname, blobresponse.get(
+                                    'versionhash')).load()
                             except botocore.exceptions.ClientError as e:
                                 if e.response['Error']['Code'] == "404":
-                                    print("Found PoDA txid! storing in db: %s" % blobresponse.get('versionhash'))
+                                    print("Found PoDA txid! storing in db: %s" %
```
Comment (sidhujag) [inline lib/poda_payload.py:128]: can you guys fix the spaces so only the differences show up.. they are all over the place and make the review needlessly hard

File: lib/poda_payload.py [hunk 2]
```diff
@@ -132,10 +168,17 @@ def get_data(self, vh):
             obj = self.s3.Object(self.bucketname, vh).get()
         except botocore.exceptions.ClientError as e:
             if e.response['Error']['Code'] == "404" or e.response['Error']['Code'] == 'NoSuchKey':
-                printdbg("Data does not exist for vh: %s" % vh)
-                return ''
+                tagData = self.storage_provider.getTagged(vh)
+                # The object does exist.
+                if (tagData.get("data") is None):
+                    printdbg("Data does not exist for vh: %s" % vh)
+                    raise
```
Comment (sidhujag) [inline lib/poda_payload.py:175]: change in functionality return '' to raise, is this intended? what are the consequences?

File: lib/poda_payload.py [hunk 3]
```diff
@@ -132,10 +168,17 @@ def get_data(self, vh):
...
+                    cid = tagData.get(
+                        "data").get("cid")
+                    data, _ = self.storage_provider.download(
+                        cid)
+                    return data.decode('utf-8')
             else:
                 # Something else has gone wrong.
                 raise
-        # The object does exist.
-        return obj['Body'].read().decode('utf-8')
```
Comment (sidhujag) [inline lib/poda_payload.py:141]: bug: this was removed? why?

File: lib/poda_payload.py (hunk 1 above)
Comment (Frank-GER) [reply lib/poda_payload.py:128]: or do separate PRs
- real code changes
- code formatting changes

**Comment by Frank-GER** [comment]
> * storage Provider Upgrade

for ...
a few words what it is needed/helpful for 

and separating real code changes from different formatting stile

File: lib/poda_payload.py (hunk 1 above)
Comment (xlassix) [reply lib/poda_payload.py:128]: @sidhujag,  Thanks for the review, The space and formatting adding weren't intended, It was done by a preset code checker [autopep8](https://pypi.org/project/autopep8/)
we would roll back to fit the initial work and disable any code checker.

File: lib/poda_payload.py (hunk 2 above)
Comment (xlassix) [reply lib/poda_payload.py:175]: @sidhujag, the Original code snippet's logic also use `raise` when the file is not found. 
kindly refer to: https://github.com/syscoin/sentinel/blob/fe952557cc8e2067857c69bb194efb4660f50986/lib/poda_payload.py#LL139C1-L139C22

I am not sure if this was done intentionally we simply mimicked the current login with the proposed change that was requested

File: lib/poda_payload.py (hunk 3 above)
Comment (xlassix) [reply lib/poda_payload.py:141]: Nice Catch!

File: lib/poda_payload.py [hunk 4]
```diff
@@ -130,12 +143,18 @@ def get_data(self, vh):
         obj = ''
         try:
             obj = self.s3.Object(self.bucketname, vh).get()
+            return obj['Body'].read().decode('utf-8')
         except botocore.exceptions.ClientError as e:
             if e.response['Error']['Code'] == "404" or e.response['Error']['Code'] == 'NoSuchKey':
-                printdbg("Data does not exist for vh: %s" % vh)
-                return ''
+                tagData = self.storage_provider.getTagged(vh)
+                # The object does exist.
```
Comment (sidhujag) [inline lib/poda_payload.py:150]: you mean doesn't exist?

File: lib/poda_payload.py [hunk 5]
```diff
@@ -130,12 +143,18 @@ def get_data(self, vh):
...
+                if (tagData.get("data") is None):
+                    printdbg("Data does not exist for vh: %s" % vh)
+                    return ''
+                else:
+                    cid = tagData.get("data").get("cid")
+                    data, _ = self.storage_provider.download(cid)
+                    return data.decode('utf-8')
             else:
                 # Something else has gone wrong.
                 raise
```
Comment (sidhujag) [inline lib/poda_payload.py:160]: should it raise or just return '', what does the API return if raise happens?

Comment (xlassix) [reply lib/poda_payload.py:160]: You are correct. It always fetches from the lighthouse if the key is not found.

recall the original code snippet's logic used `raise.`
kindly refer to: https://github.com/syscoin/sentinel/blob/fe952557cc8e2067857c69bb194efb4660f50986/lib/poda_payload.py#LL139C1-L139C22

We simply mimicked the current logic with the proposed changes that were requested. We would modify the code as required

File: lib/poda_payload.py (hunk 1 above)
Comment (xlassix) [reply lib/poda_payload.py]: Thanks, We pulled them all out

**Comment by Frank-GER** [comment]
pls fix upload issue (see above)

**Comment by Frank-GER** [comment]
Fixes check if storage to lighthouse was successful

Still only saves to lighthouse if it has to save to our PoDA storage. It needs to be higher in the hierarchy, same level as the logic for PoDA storage while not depending on it (same level).

**Comment by Frank-GER** [comment]
Let me rephrase:
It should _upload to both independently_ and can lookup each other in case the other fails

**Comment by xlassix** [comment]
@Frank-GER would you rather we use `futures` library or `threading` library to accomplish this

**Comment by Frank-GER** [comment]
Between those 2 rather the threading lib, but it can be done even sequential. Which one works best for you?
Right now it only stores to lighthouse if it has to store to our storage too.
Point is, the upload needs to be independent between the two, so if 1 already has it  the other one is still gets it if missing and the other way around.
[1] check if our storage needs it and save if necessary (orig code)
[2] check if lighthouse storage needs it and save if necessary (should not depend on [1])

**Comment by xlassix** [comment]
Thanks for the review, we made the modifications you requested

**Comment by Frank-GER** [comment]
Looks like that should do it.

**Review by Frank-GER — APPROVED** [review]
storage logic looks fixed now

**Comment by Frank-GER** [comment]
'skip lighthouse when lh_token isn't set' looks covered not too
